---
title: "Plot and Table Examples in Markdown"
author: "James Long"
always_allow_html: yes
date: "`r Sys.Date()`"
output:
  html_document:
    theme: united
    toc: yes
    toc_collapsed: yes
    toc_float: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

## colorizes text when using html or pdf output
## just spits out text x for word
colorize <- function(x, color) {
  if(color=="todo"){
    color <- "red"
    x <- paste0("TODO: ",x)
  }
  if (knitr::is_latex_output()) {
    sprintf("\\textcolor{%s}{%s}", color, x)
  } else if (knitr::is_html_output()) {
    sprintf("<span style='color: %s;'>%s</span>", color, 
      x)
  } else x
}

todo <- function(x){
  paste0("**",colorize(x,"todo"),"**")
}

library(rvg)
library(officer)
```

## Latex in Plots

```{r latex-in-plots}
library(latex2exp)
library(ggplot2)

dat <- data.frame(beta1=rnorm(100),beta2=rnorm(100))
ggplot(dat,aes(x=beta1,y=beta2)) + 
  geom_point() +
  labs(x=TeX("$\\widehat{\\beta}_1$"),
       y=TeX("$\\widehat{\\beta}_2$"))
```

## Saving Plots to pptx

```{r save-plot}
library(ggplot2)
library(rvg)
library(officer)

p <- ggplot(cars,aes(x=speed,y=dist)) +
  geom_point()
doc <- read_pptx()
doc <- add_slide(doc, layout = "Title and Content", master = "Office Theme")
my_vec_graph <- dml(code=plot(p))
doc <- ph_with(doc, my_vec_graph, location = ph_location_fullsize() )
print(doc, target = "plots.pptx")
```


## Grid of ggplots

Suggest using `cowplot` function `plot_grid` for grid of ggplots because can align axes nicely. I do not see a benefit of gridExtra.

```{r}
library(cowplot)
library(gridExtra)
p1 <- ggplot(cars,aes(x=speed,y=dist)) +
  geom_point()
p2 <- ggplot(cars,aes(x=dist,y=speed)) +
  geom_point()
p <- list(p1,p2,p2,p1)
## does not look nice because axis are not vertically aligned
##grid.arrange(grobs=p,ncol=2)
## looks much better
plot_grid(plotlist=p, ncol=2, align="v")
```


```{r grid}
## example of gridExtra, but suggest using cowplot (above) instead
library(ggplot2)
library(gridExtra)

p1 <- ggplot(cars,aes(x=speed,y=dist)) +
  geom_point()
p2 <- ggplot(cars,aes(x=dist,y=speed)) +
  geom_point()
grid.arrange(p1,p2,ncol=2)

## specify plots as list, nice when there are a lot of plots
p <- list(p1,p2)
grid.arrange(grobs=p,ncol=2)
```



## Kaplan-Meier Plots

Typically make Kaplan-Meier survival curves using `survfit` function in package survival. Plotting options:


* Preferred method: `ggsurvplot` from `survminer`. Can make a number at risk table below plot, which is helpful for understanding uncertainty.
* Backup method: `ggsurv` from `GGally` package
* Other options (examples not shown here):
    * R base graphics `plot` function. All colors and axis labels must be written manually. Requires a lot of code.
    * `autoplot` in ggfortify. Somewhat a pain to use with multiple covariates [https://stackoverflow.com/questions/50110752/ggfortify-doesnt-support-survfit-multiple-covariables](https://stackoverflow.com/questions/50110752/ggfortify-doesnt-support-survfit-multiple-covariables)


### ggsurvplot

Now use `ggsurvplot` to include number at risk table. The object returned by `ggsurvplot` is a list with 2 ggplots. So we can modify each of these ggplots to change axis labels, legend, etc.

```{r ggsurvplot-simple}
library(survminer)
library(survival)
fit  <- survfit(Surv(futime,death)~1,data=mgus)
p <- ggsurvplot(fit,data=mgus,risk.table=TRUE,ylim=c(0,1))
class(p[[1]])
p[[1]] <- p[[1]] + theme(legend.position="none")+ ylab("Survival XXX")
print(p)
```

Break down by gender and make group category labels look nice:

```{r ggsurvplot}
library(survminer)
fit  <- survfit(Surv(futime,death) ~sex, 
              data=mgus)
p <- ggsurvplot(fit,data=mgus,risk.table=TRUE,ylim=c(0,1),
                legend.title="",legend.labs=levels(mgus$sex))
print(p)
```


When zooming in, need to redefine tick mark locations on x-axis. Also change theme to get nice grid lines.

```{r ggsurvplot2}
library(survminer)
library(survival)
library(ggplot2)
fit  <- survfit(Surv(futime,death)~sex, 
              data=mgus)
p <- ggsurvplot(fit,data=mgus,risk.table=TRUE,ylim=c(0,1),xlim=c(0,5000),
                break.x.by=1000,xlab="Time (days)",
                legend.title="Gender",legend.labs=levels(mgus$sex),pval=TRUE)
p$plot <- p$plot + theme_grey() + theme(legend.position = c(0.8, 0.8))
print(p)
```

Now only plot survival curve, not at risk table

```{r}
plot(p$plot)
```

### ggsurv

```{r km-ggally}
library(survival)
library(GGally)

## one curve
mgus$old <- mgus$dxyr > median(mgus$dxyr)
fit  <- survfit(Surv(futime,death) ~1, 
              data=mgus)
p <- ggsurv(fit)
plot(p)
```


```{r km-single}
## single covariate
fit  <- survfit(Surv(futime,death) ~sex, 
              data=mgus)
p <- ggsurv(fit)
p
#p + scale_color_discrete(name="sex")
```

```{r km-multiple}
## multiple covariates, only use color to distinguish
fit  <- survfit(Surv(futime,death) ~sex+old, 
              data=mgus)
p <- ggsurv(fit)
p
## note: changing legend name is tricky because both colour and linetype must be specified as Groups
p + coord_cartesian(xlim=c(0,5000)) + labs(colour='Groups',linetype='Groups') + 
  theme(legend.position = c(0.2, 0.4),legend.title=element_blank())
```


### Competing Risk Models

`r todo("add examples with competing risk models. see fjord/2021-01-chung-local/LocalControl")`

## pairs Plots

```{r pairs}
library(ggplot2)
library(GGally)

head(iris)

## simple
p <- ggpairs(iris,columns=1:4)
p

## make points smaller
p <- ggpairs(iris,
             lower=list(continuous=wrap("points",alpha=0.3,size=0.1)),
             columns=1:4)
p

## add color
p <- ggpairs(iris,
             mapping=aes(color=Species),
             lower=list(continuous=wrap("points",alpha=0.5,size=0.5)),
             columns=1:4)
p
```

## Dotplots with ggplot

```{r dotplot}
library(ggplot2)
head(iris)

## simple dotplot
p <- ggplot(data=iris,aes(y=Sepal.Length,x=Species)) + 
      geom_dotplot(binaxis='y',
                   position=position_dodge(0.7),
                   stackdir="center",dotsize=.7,
                   binwidth=diff(range(iris$Sepal.Length)/50)) +
      labs(x=paste0("Species of Flower"),
           y=paste0("Sepal Length"))
print(p)

## add medians
p <- p + stat_summary(fun=median,geom="point",
                      aes(group=Species),
                      color="red",size=2,shape=4,show.legend = FALSE)
print(p)
```

```{r dotplot-with-fill}
## dotplots will fill and medians
mtcars2 <- mtcars
mtcars2$cyl <- as.factor(mtcars2$cyl)
mtcars2$am <- as.factor(mtcars2$am)
p <- ggplot(data=mtcars2,aes(y=mpg,x=cyl,fill=am,color=am)) +
      geom_dotplot(binaxis='y',
                   position=position_dodge(0.7),
                   stackdir="center",dotsize=.7,
                   binwidth=diff(range(mtcars2$mpg)/50))
p <- p + stat_summary(fun=median,geom="point",
                          aes(group=am),
                          position=position_dodge(.7),
                          color="black",size=2,shape=4,show.legend=FALSE)
print(p)
```


## Boxplots

Single boxplot
```{r}
library(ggpubr)
p <- ggboxplot(iris,x="Species",y="Sepal.Length",color="Species", 
               legend="none",add="dotplot",title="Sepal Length",xlab="")
print(p)
```

Set of boxplots with pvalues
```{r}
library(ggpubr)
my_comparisons <- list(c("versicolor","virginica"),
                       c("setosa","versicolor"),
                       c("setosa","virginica"))
targets <- colnames(iris)[1:4]
ps <- vector("list",length=length(targets))
for(ii in 1:length(targets)){
  ps[[ii]] <- ggboxplot(iris,x="Species",y=targets[ii],color="Species", 
                        legend="none",add="dotplot",title=targets[ii],xlab="")
  ps[[ii]] <- ps[[ii]] + stat_compare_means(comparisons=my_comparisons,method="t.test") +
    scale_y_continuous(expand = expansion(mult = c(0.05, 0.15)))
  
}
p <- cowplot::plot_grid(plotlist=ps, ncol=2, align="v")
print(p)
```

## Barplots with ggplot2

Discrete random variables can be displayed with a barplot rather than a histogram. Nice to break color bars by categorical variable to observe difference in discrete varible mean across levels of variable.

```{r barplot}
dat <- data.frame(rating=rep(c(1,3),each=50)+sample(1:4,100,replace=TRUE),
                  reviewer=rep(c("R1","R2"),each=50))
library(ggplot2)
p2 <- ggplot(dat,aes(x=rating)) +
        geom_bar()
print(p2)


p2 <- ggplot(dat,aes(x=rating,color=reviewer,fill=reviewer)) +
        geom_bar()
print(p2)
```


## ROC Curves

```{r}
library(pROC)
library(ggplot2)

## ROC Curves
true_class <- c(rep(0,50),rep(1,50))
pred1 <- rnorm(100,mean=true_class)
pred2 <- rnorm(100,mean=2*true_class)


rocobj1 <- roc(true_class,pred1)
rocobj2 <- roc(true_class,pred2)
pred_names <- c("Pred1","Pred2")
touse <- list(rocobj1,rocobj2)
names(touse) <- pred_names
g2 <- ggroc(touse)
g2 <- g2 + xlab("Specificity (# Predicted S / # True S)") + ylab("Sensitivity (# Predicted U / # True U)")

aucs <- rep(NA_real_,length(pred_names))
for(ii in 1:length(aucs)){
  aucs[ii] <- as.numeric(auc(touse[[ii]]))
}

labnames <- paste0(pred_names," (",paste0(rep("AUC=",length(aucs)),round(aucs,3)),")")
g2 <- g2 + scale_color_discrete(name = "name", labels = labnames)
g2 <- g2 + theme(legend.position = c(0.6, 0.2),legend.title=element_blank())
print(g2)
```


## Volcano Plots

```{r}
## simulate data (in log2 space)
## for performing differential expression
X <- matrix(rnorm(25*500),ncol=500)
Y <- cbind(matrix(rnorm(25*50,mean=1),ncol=50),
           matrix(rnorm(25*450),ncol=450))
dat <- rbind(X,Y)
colnames(dat) <- paste0("gene",1:ncol(dat))
group <- c(rep("groupX",25),rep("groupY",25))
```


```{r}
## determine p-values and fold changes
## assumes data is in log2 space
res <- data.frame("gene"=colnames(dat),
                  "pvalue"=rep(NA_real_,ncol(dat)),
                  "log2foldChange"=rep(NA_real_,ncol(dat)))
for(ii in 1:ncol(dat)){
  temp <- dat[,ii,drop=TRUE]
  res[ii,2] <- t.test(temp~group,var.equal=TRUE)$p.value
  res[ii,3] <- mean(temp[group=="groupX"]) - mean(temp[group=="groupY"])
}

res$padj <- p.adjust(res$pvalue,method="fdr")
```


```{r}
res$foldChange <- 2^res$log2foldChange
library(ggplot2)
library(ggrepel)
library(scales)

## make plot
res$delabel <- NA_character_
ix <- res$padj < 0.05
ix[is.na(ix)] <- FALSE
res$delabel[ix] <- res$gene[ix]
p <- ggplot(res,aes(x=foldChange,y=padj,label=delabel)) +
  geom_point() +
  geom_hline(yintercept=0.05, col="red") + 
  xlab("Fold Change") + 
  geom_text_repel(size=2) +
  scale_x_continuous(trans='log2') + 
  scale_y_continuous(trans=c("log10","reverse")) +
  ylab("adjusted pvalue")
print(p)
```

## log axes

```{r}
library(ggplot2)
n <- 1000
df <- data.frame(x=abs(rt(n,df=1)),y=abs(rt(n,df=1)))
ggplot(df,aes(x=x,y=y)) + 
  geom_point()

ggplot(df,aes(x=x,y=y)) + 
  geom_point() +
  scale_y_continuous(trans='log10') + scale_x_continuous(trans='log10')

## suppress scientific notation
options(scipen=999)
ggplot(df,aes(x=x,y=y)) + 
  geom_point() +
  scale_y_continuous(trans='log10') + scale_x_continuous(trans='log10')

ggplot(df,aes(x=x,y=y)) + 
  geom_point() +
  coord_trans(x="log10", y="log10")

```

## ggplot2 Notes

ggplot2 language

* aesthetic: x, y, color, shape, etc
* scale: each plot aesthetic is association with 1 scale which is a function from data space to aesthetic space
* guides: legends and axes. allow user to map values from plot back to data
* layer: collection of geometric objects (geoms)

useful functions

* labs: specify name associated with scale




```{r practice}
toy <- data.frame(
  const = 1, 
  up = 1:4,
  txt = letters[1:4], 
  big = (1:4)*1000,
  log = c(2, 5, 10, 2000)
)
base <- ggplot(toy, aes(const, up)) +
  scale_x_continuous(NULL, breaks = NULL)
base + geom_point(aes(colour = txt))
base + geom_point(aes(shape = txt))
base + geom_point(aes(shape = txt, colour = txt))

base <- ggplot(toy, aes(const, up))

base + geom_point(aes(colour = txt))
?scale_x_continuous
```


## gtsummary

[gtsummary](https://cran.r-project.org/web/packages/gtsummary/index.html) makes tables to summarize models coefficient fits and variable distributions.

```{r summary-table}
library(gtsummary)
a <- tbl_summary(cars[1:10,])
knit_print(a)

add_p_ex1 <-
  trial[c("age", "grade", "trt")] %>%
  tbl_summary(by = trt) %>%
  add_p()

add_p_ex1
```

Contingency table with gtsummary package:
```{r table-table}
library(gtsummary)
dat <- data.frame("M1"=c("Correct","Correct","Correct","Correct","Incorrect","Incorrect"),
                  "M2"=c("Correct","Correct","Correct","Incorrect","Incorrect","Incorrect"))
tbl_cross(dat)
```


## kableExtra

[kableExtra](https://cran.r-project.org/web/packages/kableExtra/vignettes/awesome_table_in_html.htmlhttps://cran.r-project.org/web/packages/kableExtra/vignettes/awesome_table_in_html.htmlhttps://cran.r-project.org/web/packages/kableExtra/vignettes/awesome_table_in_html.htmlhttps://cran.r-project.org/web/packages/kableExtra/vignettes/awesome_table_in_html.html) makes nice tables.

### Basic Tables

Simple Table
```{r simple-table}
library(kableExtra)
head(cars)
kbl(cars[1:10,]) %>%
  kable_paper(bootstrap_options="striped",full_width = F)
```

Nice header joining columns:
```{r header-table}
library(kableExtra)
head(iris)
iris2 <- iris[1:10,c(5,1:4)]
kbl(iris2)%>%
  kable_paper(bootstrap_options="striped",full_width = F) %>%
  add_header_above(c(" "=1,"Sepal"=2,"Petal"=2))
```




### Scrolling Tables

```{r scroll-table}
library(kableExtra)
head(cars)
kbl(cars)%>%
  kable_paper(bootstrap_options="striped",full_width = F) %>%
  scroll_box(height="200px",width="40%")
```


```{r scroll-table-missing}
library(kableExtra)
dat <- cars
dat[sample(1:nrow(dat),10),1] <- NA

num_na <- apply(dat,2,function(x){sum(is.na(x))})
num_na <- data.frame("number missing"=num_na,
                     "fraction missing"=round(num_na/nrow(dat),2))
num_na <- num_na[order(num_na[,1],decreasing=TRUE),]
kbl(num_na)%>%
  kable_paper(bootstrap_options="striped",full_width = F) %>%
  scroll_box(height="200px",width="40%")
```





## Table Output

* Input Formats: 1) data.frame 2) gtsummary
* Output formats: 1) html 2) latex 3) word

We summarize the recommended functions / packages for handling these each setting.

```{r}
library(kableExtra)
input <- c("data.frame","","","gtsummary","","")
output <- rep(c("html","latex","word"),2)
method <- c("kableExtra","kableExtra","flextable","---","huxtable","flextable")
df <- data.frame(input=input,output=output,method=method)
kbl(df) %>%
  kable_paper(bootstrap_options="striped",full_width = F)
```



### data.frame

html output
```{r}
library(kableExtra)
head(cars)
tab <- kbl(cars[1:10,]) %>%
        kable_paper(bootstrap_options="striped",full_width = F)
tab ## directly output to html
```

tex output
```{r}
## to 
## good table advice https://haozhu233.github.io/kableExtra/awesome_table_in_pdf.pdf
library(kableExtra)
iris <- iris[,c(5,1:4)]
caption <- "Iris data set  \\label{tab:iris}"
tab <- kbl(iris[1:20,],booktabs=TRUE,format="latex",escape=FALSE,caption=caption) %>%
    add_header_above(c(" "=1,"Sepal"=2,"Petal"=2))
save_kable(tab,file="iris.tex")
```


word output
```{r}
library(flextable)
ft <- flextable(cars[1:10,])
ft <- width(ft, width = 0.6)
ft <- height(ft,height=0.25)
ft <- hrule(ft, rule = "exact")
save_as_docx(ft,path="cars.docx")
```

### gtsummary


html

```{r}
library(gtsummary)
add_p_ex1 <-
  trial[c("age", "grade", "trt")] %>%
  tbl_summary(by = trt) %>%
  add_p()

add_p_ex1

```


tex

```{r}
library(gtsummary)
add_p_ex1 <-
  trial[c("age", "grade", "trt")] %>%
  tbl_summary(by = trt) %>%
  add_p()
ht <- as_hux_table(add_p_ex1)
huxtable::bottom_padding(ht) <- 0
huxtable::top_padding(ht) <- 0
huxtable::right_padding(ht) <- 0
huxtable::caption(ht) <- "Huxtable example."
huxtable::label(ht) <- "tab:table1-within5"

## convert to latex
ht_latex <- ht %>% huxtable::to_latex()
## remove some formatting
ht_latex <- gsub("```{=latex}","",ht_latex,fixed=TRUE)
ht_latex <- gsub("```","",ht_latex,fixed=TRUE)
cat(ht_latex,file="huxtable-example.tex")

## NOTE: run the below code and add these packages to latex file preamble
## huxtable::report_latex_dependencies(quiet = FALSE, as_string = FALSE)
```



word: To reduce size of cells in word, right click on tables, table properties, row, and then specify 0.25in exact height.

```{r word-table}
library(flextable)
library(gtsummary)
add_p_ex1 <-
  trial[c("age", "grade", "trt")] %>%
  tbl_summary(by = trt) %>%
  add_p()
ft <- as_flex_table(add_p_ex1)
#ft <- width(ft, width = 0.6)
ft <- height(ft,height=0.25)
ft <- hrule(ft, rule = "exact")
save_as_docx(ft,path="table1.docx")
```
